src/stac_task/payload.py

```diff
@@ -0,0 +1,110 @@
+from .task import Input, Output, Task
+from .types import PathLikeObject
+
+
+class Payload(BaseModel):
+    """A payload describing the items and the tasks to be executed
+
+    Pretty specific to [cirrus](https://github.com/cirrus-geo/)."""
+
+    type: Literal["FeatureCollection"] = "FeatureCollection"
+    """Must be FeatureCollection."""
+
+    features: List[Dict[str, Any]] = []
+    """A list of STAC items, or things sort of like STAC items."""
+
+    process: Process = Process()
+    """The process definition."""
+
+    # TODO do we need to support `url` as well?
+    href: Optional[str] = None
+    """An optional href parameter, used in indirect payloads.
+    
+    Indirect payloads contain an href to a large payload living on s3.
+    """
+
+    self_href: Optional[str] = Field(default=None, exclude=True)
+    """The location that the payload was read from.
+    
+    Used to resolve relative hrefs.
+    """
+
+    @classmethod
+    def from_href(cls, href: str, allow_indrections: bool = True) -> Payload:
+        """Loads a payload from an href.
+
+        If the payload has an `href` attribute set, that href will be fetched.
+        This is used for "indirect" payloads that point to large payloads that
+        need to be stored on s3.
+
+        Args:
+            href: The href to load the payload from.
+            allow_indirections: Whether to follow indirection links. Generally
+                used only when recursively calling this function to prevent infinite
+                indirection loops.
+
+        Returns:
+            Payload: The payload
+        """
+        # TODO we could go async with these
+        payload = cls.model_validate_json(stac_asset.blocking.read_href(href))
+        if payload.href and not payload.features:
+            if allow_indrections:
+                href = pystac.utils.make_absolute_href(
+                    payload.href, href, start_is_dir=False
+                )
+                return cls.from_href(href, allow_indrections=False)
+            else:
+                raise ValueError("Multiple indirections are not supported")
+        else:
+            payload.self_href = href
+            return payload
+
+    def execute(self, tasks: Mapping[str, Type[Task[Input, Output]]]) -> Payload:
+        """Executes this payload, returning the updated payload.
+
+        Args:
+            tasks: The tasks to use for processing.
+
+        Returns:
+            Payload: A new payload, with updated items.
+        """
+        matches = set(key for key in tasks.keys() if key in self.process.tasks)
+        if len(matches) == 0:
+            raise ExecutionError("no tasks to execute")
+        elif len(matches) == 1:
+            key = matches.pop()
+            task_class = tasks[key]
+            task = task_class(**self.process.tasks[key])
+            task.payload_href = self.self_href
+            items = list()
+            # TODO async / thread pool
+            for item in self.features:
```

Comment:
A task doesn't work like this.  A process function takes in all the Items from the input and uses all of them to do something.  
Although I suppose this could be an additional feature that could add in another way to batch jobs...but that raises issues around state tracking and exception handling for batches I think are best avoided.

Reply:
I was thinking about things from the developer's perspective, where each incoming item/href/assets is independent of the others. When I'm writing a `process` method, I only want to be thinking about one item at a time, right?

The handling of partial failures is best left to **stac-task**, which can collect and partition successful and unsuccessful calls to process. [We do something similar in **stac-asset**](https://github.com/stac-utils/stac-asset/blob/6c8f6c3a924fe51ba1e0c1d9d41d0b4195b6a577/src/stac_asset/_functions.py#L150-L165):

```python
        exceptions = list()
        for result in results:
            if isinstance(result, WrappedError):
                if self.config.error_strategy == ErrorStrategy.DELETE:
                    del result.download.owner.assets[result.download.key]
                else:
                    # Simple check to make sure we haven't added other
                    # strategies that we're not handling
                    assert self.config.error_strategy == ErrorStrategy.KEEP


                if self.config.warn:
                    warnings.warn(str(result.error), DownloadWarning)
                else:
                    exceptions.append(result.error)
        if exceptions:
            raise DownloadError(exceptions)
```

Reply:
No, a process function is using all of the Items to create it's outputs.  Consider a data fusion example, where we are combining Landsat and Sentinel into a single derived product, or pan-sharpening.

Cirrus, and SWOOP is based on the idea of a single job having a single entry in the database. If you batch within a task then your state db does not represent a single job, it represents X jobs.  If Y of those fail and you want to rerun the others then those would end up going into the statedb leading to a mix of "batch" tasks and single item tasks.

In short, the orchestration platform is responsible for batching and orchestration, it should not be left to the tasks, but interested in @jkeifer's thoughts here.

Reply:
> No, a process function is using all of the Items to create it's outputs.

Still think this is an anti-pattern, because you can't define a schema for the `features` attribute of a payload, since it's a list. IMO the upstream "gathering" task should perform search-and-match, generating one item for each set of inputs that you want to fuse, e.g.:

```json
"features": [
  {
    "id": "item-0",
    "properties": {
       "start_datetime": "2023-09-19T00:00:00Z",
       "end_datetime": "2023-09-20T00:00:00Z"
    },
    "assets": {
       "landsat_band": { "you-get": "the-idea" },
       "sentinel_2_band": { }
    }
  },
  {
    "id": "item-1",
    "properties": {
       "start_datetime": "2023-09-18T00:00:00Z",
       "end_datetime": "2023-09-19T00:00:00Z"
    },
    "assets": {
      "landsat_band": {},
      "sentinel2_band": {}
    }
  }
]
```

That being said, the current architecture can easily switch to support your model. I'd just change the line we're commenting on to be:

```python
task.process_all(self.features)
```

and define a default `process_all` that does:

```python
def process_all(self, features: List[Input]) -> List[Output]:
    if len(features) != 1:
        raise NotImplemented("Cannot process of a list of more than one feature")
    else:
        return self.process(features[0])
```

Then, if tasks _did_ want to work on a list of (unschemad) inputs, they could just override `process_all`.

Reply:
It certainly makes data provenance a real problem, and puts a lot more logic in gathering together a payload into the feeder, it's pretty different then the original idea of how STAC workflows would work.  🤷

Reply:
> puts a lot more logic in gathering together a payload into the feeder, it's pretty different then the original idea of how STAC workflows would work. 

Logic's got to go somewhere, and IMO its better placed closer to the initial search, rather than separating the two.

> it's pretty different then the original idea of how STAC workflows would work. 

Fair enough, I am admittedly coming at this from a non-Cirrus perspective. Again, there's not much we'd have to change technically to support the system you describe, so we're not far off I don't think.

Reply:
Okay, I'm going to resolve this thread, since we now have `StacOutTask` and `StacInStacOutTask` to directly replicate the original task model.